### DISCUSSION TITLE
Keymaps for Codeium

### DISCUSSION THREAD

**Discussion by UnclePsymon** [nix-community/nixvim#1872] [Q&A]
Hey there! I want to set the keybinds for Codeium, but i can't get them to work.

This is from my old lua config:
```lua
vim.g.codeium_disable_bindings = 1
vim.keymap.set('i', '<C-u>', function () return vim.fn['codeium#Accept']() end, { expr = true })
vim.keymap.set('i', '<c-h>', function() return vim.fn['codeium#CycleCompletions'](1) end, { expr = true })
vim.keymap.set('i', '<c-l>', function() return vim.fn['codeium#CycleCompletions'](-1) end, { expr = true })
vim.keymap.set('i', '<c-k>', function() return vim.fn['codeium#Clear']() end, { expr = true })
```
I tried it like that, but that did not work:
```nix
keymaps = [
  {
    action = "<cmd>codeium#CycleCompletitions<CR>";
    key = "<C-u>";
    mode = "i";
    options = { 
      desc = "Codeium Cycle Completitions";
      expr = true;
    };
];
```

Can you tell me how to do it? ❤️

**Comment by MattSturgeon** [comment]
In your first example you are defining lua functions as your rhs (action), while in your nixvim example you are not.

Probably better to port what you had before exactly before attempting to optimise it:

```nix
keymaps = [
  {
    action.__raw = "function() return vim.fn['codeium#Accept']() end";
    key = "<C-u>";
    mode = "i";
    options = { 
      desc = "Codeium Cycle Completitions";
      expr = true;
    };
  }
];
```

I think `action.__raw = "vim.fn['codeium#Accept']"` would also work.

---

IDK if the lua function wrapping a vim function is strictly necessary, it should be possible to use the vim fn directly as you proposed `<cmd>codeium#CycleCompletitions<CR>`, but better to try one thing at a time.

Have you tried calling the function directly from vim's command mode (without a keymapping)?

```
:codeium#CycleCompletitions
```

---

The only other question is; is codeium actually enabled and loaded?

**Reply by UnclePsymon** [reply]
Thanks for the reply!
Codeium is enabled and also shows me suggestions under my cursor (i think this is nvim-cmp).  `action.__raw = "function() return vim.fn['codeium#Accept']() end";` and `action.__raw = "vim.fn['codeium#Accept']"` both give me `unknown funktion` errors. `:codeium#CycleCompletitions` and `:Codeium#CycleCompletitions` do not exist.

**Reply by MattSturgeon** [reply]
Looking upstream, [their example](https://github.com/Exafunction/codeium.vim?tab=readme-ov-file#%EF%B8%8F-keybindings) is

```vim
imap <script><silent><nowait><expr> <C-g> codeium#Accept()
imap <C-;>   <Cmd>call codeium#CycleCompletions(1)<CR>
imap <C-,>   <Cmd>call codeium#CycleCompletions(-1)<CR>
imap <C-x>   <Cmd>call codeium#Clear()<CR>
```

So to use the `<cmd>` style you also need to use the `call` command.

Have you tried `:call codeium#CycleCompletitions(1)`?

---

However, if you're getting "unknown function" when doing `vim.fn['codeium#Accept']` then something else is not right.

**Reply by UnclePsymon** [reply]
`:call codeium#CycleCompletitions(1)`  gives me unknown funktion. 
I just managed to use `cmp` to use the completitions. This is actually more convinient for me, so I'll use that. Thanks for the help :)
Unless you need me for some bug hunting because of the "unknown function"?

**Reply by MattSturgeon** [reply]
Glad you found a solution. I'd be surprised if there's an issue on our end, so I'm happy to close this.

Thanks for reporting back in!
